File: automation_testing/ClickHouse/dotnet-client/AsabruCH.cs
using System.Data;
using ClickHouse.Ado;

class AsabruCH
{
    private readonly string CONFIG_FILE = "../ch_config.yaml";
    private readonly string SQL_FILE = "../ch_sql.yaml";

    private string _chHost = string.Empty;
    private string _chUser = string.Empty;
    private string _chPassword = string.Empty;
    private string _datasetUrl = string.Empty;
    private string _datasetFileName = string.Empty;
    private string _datasetCsvName = string.Empty;
    private string _databaseName = string.Empty;
    private string _tableName = string.Empty;
    private int _port = 0;
    private string _caCert = string.Empty;
    private string _clientConfig = string.Empty;
    private bool _ssl = false;

    private Dictionary<string, Dictionary<string, string>> _sqlStatements = new Dictionary<string, Dictionary<string, string>>();

    public AsabruCH()
    {
        // Load configuration from YAML file
        LoadConfiguration();
        // Load SQL statements from YAML file
        LoadSqlStatements();
    }

    private void LoadConfiguration()
    {
        Console.WriteLine("Started : LoadConfiguration");
        // Load configuration from YAML file
        var config = new YamlDotNet.Serialization.Deserializer().Deserialize<Dictionary<string, string>>(
            File.ReadAllText(CONFIG_FILE));

        _chHost = config["host"];
        _chUser = config["username"];
        _chPassword = config["password"];
        _datasetUrl = config["dataset_url"];
        _datasetFileName = config["dataset_file_name"];
        _datasetCsvName = config["dataset_csv_name"];
        _databaseName = config["database"];
        _tableName = config["table"];
        _port = int.Parse(config["port"]);
        _caCert = config["ca_cert"];
        _clientConfig = config["client_config"];
        _ssl = bool.Parse(config["ssl"].ToLower());

        if (_ssl)
        {
            Console.WriteLine("Using SSL/TLS");
        }
        Console.WriteLine("Finished : LoadConfiguration");
    }

    private void LoadSqlStatements()
    {
        _sqlStatements = new YamlDotNet.Serialization.Deserializer().Deserialize<Dictionary<string, Dictionary<string, string>>>(
            File.ReadAllText(SQL_FILE));
    }



    private readonly string CommonConnectionString = "Compress=True;CheckCompressedHash=False;SocketTimeout=60000;";
    private ClickHouseConnection CreateConnection()
    {
        var sslConnectionString = _ssl
            ? $"SslMode=VerifyFull;SslCa={_caCert};"
            : string.Empty;

        var connectionString = $"{CommonConnectionString}Host={_chHost};Port={_port};User={_chUser};Password={_chPassword};{sslConnectionString}";

        Console.WriteLine("Connection String : " + connectionString);
        return new ClickHouseConnection(connectionString);
    }

    public void BenchmarkPerformanceOld()
    {
        var loopStart = DateTime.Now;
        var loopTimes = new List<double>();
        var iters = 10;

        Console.WriteLine("Starting the benchmark");
        using (var connection = CreateConnection())
        {
            connection.Open();
            for (var x = 0; x < iters; x++)
            {
                Console.WriteLine($"Iteration {x + 1}");

                var iterStart = DateTime.Now;

                foreach (var queryKey in _sqlStatements["read"].Keys)
                {
                    var query = _sqlStatements["read"][queryKey];
                    connection.CreateCommand(query).ExecuteNonQuery();
                    Console.WriteLine(query);
                }

                var iterEnd = (DateTime.Now - iterStart).TotalSeconds;

                loopTimes.Add(iterEnd);
            }
            connection.Close();
        }

        Console.WriteLine($"Average Time taken: {loopTimes.Average()} seconds");
        Console.WriteLine($"Median Time taken: {loopTimes[loopTimes.Count / 2]} seconds");
    }

    public void BenchmarkPerformance()
    {
        var loopStart = DateTime.Now;
        var loopTimes = new List<double>();
        var iters = 10;

        Console.WriteLine("Starting the benchmark");

        using (var connection = CreateConnection())
        {
            connection.Open();

            for (var x = 0; x < iters; x++)
            {
                Console.WriteLine($"Iteration {x + 1}");

                var iterStart = DateTime.Now;

                foreach (var queryKey in _sqlStatements["read"].Keys)
                {
                    var query = _sqlStatements["read"][queryKey];

                    using (var cmd = new ClickHouseCommand(connection, query))
                    {
                        cmd.ExecuteNonQuery();
                    }

                    Console.WriteLine($"Query: {query}, Time: {(DateTime.Now - iterStart).TotalSeconds} seconds");
                }

                var iterEnd = (DateTime.Now - iterStart).TotalSeconds;

                loopTimes.Add(iterEnd);
            }

            connection.Close();
        }

        Console.WriteLine($"Average Time taken: {loopTimes.Average()} seconds");
        Console.WriteLine($"Median Time taken: {loopTimes[loopTimes.Count / 2]} seconds");
    }

}

File: automation_testing/ClickHouse/python-client/AsabruCH.py
from clickhouse_driver import Client
import os
import yaml
import requests
import lzma
import time
from statistics import mean, median


class AsabruCH:
    def __init__(self, config_file='../ch_config.yaml', sql_file = '../ch_sql.yaml') -> None:
        
        with open(config_file, "r") as f:
            config = yaml.safe_load(f)

        with open(sql_file, "r") as f:
            self.sql_statements = yaml.safe_load(f)

        self.ch_host = config['host']
        self.ch_user = config['username']
        self.ch_password = config['password']
        self.dataset_url = config['dataset_url'] 
        self.dataset_file_name = config['dataset_file_name']
        self.dataset_csv_name = config['dataset_csv_name']
        self.database_name = config['database']
        self.table_name = config['table']
        self.port = config['port']
        self.ca_cert = config['ca_cert']
        self.client_config = config['client_config']
        self.ssl = True if config['ssl']=='True' else False

        if self.ssl:
            print('Using SSL/TLS')


    def create_client(self):
        if not self.ssl:
            client = Client(host=self.ch_host, user=self.ch_user, password=self.ch_password, port=self.port, 
                                send_receive_timeout=600)
        if self.ssl:
            client = Client(host=self.ch_host, user=self.ch_user, password=self.ch_password, port=self.port, 
                                send_receive_timeout=600, ca_certs=self.ca_cert, secure=True )

        return client

    def benchmark_performance(self):
        loop_start = time.time()
        loop_times = []

        iters = 10
        queries = len(self.sql_statements['read'].keys())

        print('Starting the benchmark')

        for x in range(iters):
            client = self.create_client()
            print(f'Iteration {x+1}')

            iter_start = time.time()

            for query_key in self.sql_statements['read'].keys():
                query = self.sql_statements['read'][query_key]
                client.execute(query)
                print(query)
            
            iter_end = time.time() - iter_start
            client.disconnect()

            loop_times.append(iter_end)

        print ('Average Time taken : ', mean(loop_times))
        print ('Median Time taken : ', median(loop_times))


        

